pkg/metrics/object_count.go

```diff
@@ -0,0 +1,111 @@
+package metrics
+
+import (
+	"context"
+
+	corev1 "github.com/rancher/rancher/pkg/generated/norman/core/v1"
+	managementv3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	projectv3 "github.com/rancher/rancher/pkg/generated/norman/project.cattle.io/v3"
+	rbacv1 "github.com/rancher/rancher/pkg/generated/norman/rbac.authorization.k8s.io/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+)
+
+type clusterObjectCounter interface {
+	ConfigMaps(context.Context) (int64, error)
+	Namespaces(context.Context) (int64, error)
+	Nodes(context.Context) (int64, error)
+	Secrets(context.Context) (int64, error)
+	ClusterRoleBindings(context.Context) (int64, error)
+	RoleBindings(context.Context) (int64, error)
+	AppRevisions(context.Context) (int64, error)
+	CatalogTemplateVersions(context.Context) (int64, error)
+	Projects(context.Context) (int64, error)
+}
+
+type clusterCountClient struct {
+	// There are *complete* caches only for some kinds, use that when available (see pkg/controllers/managementuser/controllers.go).
+
+	core interface {
+		corev1.NamespacesGetter
+	}
+	rbac interface {
+		rbacv1.ClusterRoleBindingsGetter
+		rbacv1.RoleBindingsGetter
+	}
+
+	// Otherwise, use a client directly.
+
+	dynamic dynamic.Interface
+}
+
+func (h *metricsHandler) getClusterClient(clusterID string) (clusterObjectCounter, error) {
+	cluster, err := h.clusterManager.UserContext(clusterID)
+	if err != nil {
+		return nil, err
+	}
+	return &clusterCountClient{
+		dynamic: dynamic.New(cluster.UnversionedClient),
+		core:    cluster.Core,
+		rbac:    cluster.RBAC,
+	}, nil
+}
+
+func (c clusterCountClient) countAllObjects(ctx context.Context, gvr schema.GroupVersionResource) (int64, error) {
+	// Starting on Kubernetes 1.16, ListMeta includes a RemainingItemCount field, with an estimation of the number of items that couldn't fit in the request
+	// Use that to obtain the number of objects without having to perform all the calls.
+	list, err := c.dynamic.Resource(gvr).List(ctx, metav1.ListOptions{Limit: 1})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return 0, nil
+		}
+		return 0, err
+	}
+	n := int64(len(list.Items))
+	if remaining := list.GetRemainingItemCount(); remaining != nil {
+		n += *remaining
+	}
```

Comment:
Shouldn't `n` always be 1 at this point?

Therefore

```suggestion
	n := 1
	if remaining := list.GetRemainingItemCount(); remaining != nil {
		n += *remaining
	}
```

Reply:
It could be 0, isn't it?

Reply:
I confirm that empty results return HTTP 200 with an empty `items` field.
```
❯ kubectl get cronjobs -A -o json -v=6
I0826 10:55:19.877275   54318 round_trippers.go:553] GET https://0.0.0.0:6445/apis/batch/v1/cronjobs?limit=500 200 OK in 10 milliseconds
{
    "apiVersion": "v1",
    "items": [],
    "kind": "List",
    "metadata": {
        "resourceVersion": ""
    }
}
```

Reply:
Then I am confused about the `apierrors.IsNotFound` check above, in that I expected such error if the list was empty. Can that happen and in what cases?